cmd/snapd-generator/main.c

```diff
@@ -30,19 +31,168 @@
 #include "../libsnap-confine-private/mountinfo.h"
 #include "../libsnap-confine-private/string-utils.h"
 
-static sc_mountinfo_entry *find_root_mountinfo(sc_mountinfo * mounts)
+static sc_mountinfo_entry *find_dir_mountinfo(sc_mountinfo *mounts, const char *mnt_dir)
 {
 	sc_mountinfo_entry *cur, *root = NULL;
 	for (cur = sc_first_mountinfo_entry(mounts); cur != NULL;
 	     cur = sc_next_mountinfo_entry(cur)) {
-		// Look for the mount info entry for the root file-system.
-		if (sc_streq("/", cur->mount_dir)) {
+		// Look for the mount info entry. We take the last one, which
+		// would be the last mount on top of mnt_dir.
+		if (sc_streq(mnt_dir, cur->mount_dir)) {
 			root = cur;
 		}
 	}
 	return root;
 }
 
+static int create_early_mount(const char *normal_dir,
+			      const char *what,
+			      const char *where)
+{
+	// Replace directory separators with dashes to build the unit name.
+	// WARNING this is limited to directories that do no contain dashes,
+	// which is fine at the moment as our mountpoints are
+	// /usr/lib/{modules,firmware}.
```

Comment:
I think we should have this warning before the function definition so that if other people plan to call this function, they are aware of the limitation. It is not obvious it is something to pay attention when it is in the code.

Reply:
Systemd already has that function, perhaps we could copy it with the right license.